datafusion/datasource/src/source.rs

```diff
@@ -174,6 +175,21 @@ impl ExecutionPlan for DataSourceExec {
         self.data_source.statistics()
     }
 
+    fn statistics_by_partition(&self) -> datafusion_common::Result<Vec<Statistics>> {
```

Comment:
The partition-level statistic of `DataSource` will be passed to downstream nodes. 

We need to implement the method for different nodes.

Reply:
We have our own in-house version of ProgressiveEval which we are currently iterating upon to make generalized. This part (extracting and ordering all partitions in the DAG below) is indeed the majority of the work thus far.

We've debugged and gotten unit tests for the majority of that code. I'll have a chat with @alamb about how much we can move upstream to the apache project.

Reply:
I think this part will work well after this PR is merged: https://github.com/apache/datafusion/pull/15432 which introduces  the partition-level statistics.

And after https://github.com/apache/datafusion/pull/15432 is merged, we can add the `statistics_by_partition` API to `ExecutionPlan` and implement it for all nodes.